### PR TITLE
storage: Fix GCBytesAge in CheckSSTConflicts with range keys

### LINKED ISSUE #82920: batcheval: `AddSSTable` does not adjust `GCBytesAge`
When `AddSSTable` checks for conflicts or shadowing via `CheckSSTConflicts`, it is supposed to accurately adjust the SST's `MVCCStats` to reflect the effect it has on the existing data. However, it makes no attempt at all to adjust `GCBytesAge`, leading to incorrect stats.

To see this, change e.g. `TestEvalAddSSTable` to use timestamps that are larger than 1 second, i.e. 1 billion nanoseconds -- so use `1e9` instead of `1` in timestamps -- and pass through the request timestamp to all stats calculations. The stats assertions will then immediately fail. This is because `GCBytesAge` computations only care about whole seconds, rounded down, so timestamps less than `1e9` will always result in 0 `GCBytesAge`.

There is some draft work in this branch: https://github.com/erikgrinaker/cockroach/tree/addsstable-stats

There are several scenarios that need handling:

* The SST shadows an existing value.
* An existing value shadows the SST value.
* A value in the SST shadows another value in the SST.
* All of the above, but with tombstones rather than values.

A fix for the simplest case is e.g.:

https://github.com/erikgrinaker/cockroach/blob/5c20fd3ea286fee34df577c2d92ef62c60e2b4fa/pkg/storage/sst.go#L197-L215

Jira issue: CRDB-16730

Epic CRDB-2624

### PR DESCRIPTION
**Pull request by itsbilal**
Change #87303 fixed GCBytesAge in `CheckSSTConflicts` for points, but not for range keys. This change fixes it for range keys too, by ensuring the changes to statsDiff happen at the right timestamp using statsDiff.AgeTo.

Fixes #82920.

Release note: None.

### CODE REVIEW

File: pkg/storage/sst.go [hunk 1]
```diff
@@ -460,12 +463,25 @@ func CheckSSTConflicts(
...
+						updatedStack.Versions = extRangeKeys.Versions.Clone()
+						for i, v := range sstRangeKeys.Versions {
+							if i == 0 {
+								// We do this dance to make updatedStack.Versions.Newest() == v. This
+								// is necessary to keep GCBytesAge calculations correct, we don't
+								// want updateStatsOnRangeKeyPutVersion to "lift" the GCBytesAge
+								// contribution of extRangeKeys' bounds. We will do that later.
+								// We only want it to add the version.
+								oldVersions := updatedStack.Versions
+								updatedStack.Versions = append(MVCCRangeKeyVersions{}, v)
```
Comment (erikgrinaker) [inline pkg/storage/sst.go:476]: nit: Can use `MVCCRangeKeyVersions{v}, oldVersions...)` to use a single allocation here. The dance is also a bit unfortunate, we can consider either inlining the logic or adding a new helper function instead.

**Review by erikgrinaker — APPROVED** [review]
Thanks for seeing this through!

**Comment by itsbilal** [comment]
TFTR!

**Comment by itsbilal** [comment]
bors r=erikgrinaker

**Pull request merged**